### DISCUSSION TITLE
Get source and hierarchy of nbgrader_config.py files

### DISCUSSION THREAD

**Discussion by joebathelt** [jupyter/nbgrader#1833] [Q&A] [answered]
Hi everyone,

After thoroughly reviewing the nbgrader and JupyterHub documentation, I successfully set up nbgrader on a new Digital Ocean instance. However, I'm unable to replicate this on our local server. I suspect that residual settings from a prior installation might be causing the issue. It seems that nbgrader might not be accessing the correct nbgrader_config.py file. 

Output from `jupyter --paths`:
```bash
config:
    /home/jmbathe/.jupyter
    /home/jmbathe/.local/etc/jupyter
    /opt/tljh/user/etc/jupyter
    /usr/local/etc/jupyter
    /etc/jupyter
data:
    /home/jmbathe/.local/share/jupyter
    /opt/tljh/user/share/jupyter
    /usr/local/share/jupyter
    /usr/share/jupyter
runtime:
    /home/jmbathe/.local/share/jupyter/runtime
```
Current minimal config file
```python
c = get_config()
c.GenerateConfigApp.logfile = "/usr/local/share/jupyter/nbgrader.log"
```

How can I determine which configuration files nbgrader reads and the sequence in which they're read? I attempted to reset them by specifying only a log file location and copied this configuration to all folder on the Jupyter path, but I couldn't even get the error log to function.

Thank you in advance!

Best,

Joe

**Comment by jeflem** [comment]
Hi @joebathelt,

nbgrader is designed as a stand-alone tool working from the command line. But nbgrader also ships with some Jupyter extensions to connect the command line only nbgrader to JupyterJub/JupyterLab. Thus, nbgrader (afaik) does not care about config paths of JupyterHub.

If used with JupyterHub there should be three `nbgrader_config.py` in different locations for different purposes. Have a look at [Use Case 3: nbgrader and JupyterHub](https://nbgrader.readthedocs.io/en/stable/configuration/nbgrader_config.html#use-case-3-nbgrader-and-jupyterhub) in nbgrader docs for details.

Best regards,

Jens

**Reply by joebathelt** [reply]
Thank you, @jeflem. The description of Case 3 is what I found puzzling. It suggests a hierarchy consisting of a global configuration file, a secondary file for determining the course's location, and a final local file specific to the course. I comprehend the positions for both the global and local files, but I'm uncertain about the middle one.

Regarding this secondary configuration file, the documentation references Case 2, specifying it's for "your main course directory (where you run the nbgrader commands)". This makes sense when operating nbgrader locally. However, I'm unclear about the directory from which nbgrader operates when on JupyterHub.

**Comment by jeflem** [comment, marked as answer]
Had a look at my setup: There are only two `nbgrader_config.py`. A global one in `/etc/jupyter/nbgrader_config.py`:
```
from nbgrader.auth import JupyterHubAuthPlugin

c = get_config()

c.Exchange.path_includes_course = True
c.Authenticator.plugin_class = JupyterHubAuthPlugin
c.Exchange.root = '/opt/nbgrader_exchange'
c.NbGrader.logfile = '/share/jupyter/nbgrader.log' 
```
The second `nbgrader_config.py` is per course. For each course there is a separate user account running the formgrader service for JupyterHub. So for course `test_course` we have in `/home/grader_test_course/.jupyter/nbgrader_config.py`:
```
c = get_config()

c.CourseDirectory.root = '/home/grader_test_course/course_data'
c.CourseDirectory.course_id = 'test_course'

# plus some lines for dealing with hidden tests in feedbacks
```
Two config files should suffice as long as we have only one course per formgrader service. Never used one formgrader for multiple courses, but seems to be possible according to the docs.

Jens

**Reply by joebathelt** [reply]
@jeflem, thank you very much! I get it now!
